### PR TITLE
[Done] Added the end points for the Profile model

### PR DESCRIPTION
**Pull request by Saurav-Shrivastav**
- Discrepancies in the update view, currently it uses the user_id (needs to be provided in the URL) to perform updates, need to fix that!

### CODE REVIEW

**Comment by Saurav-Shrivastav** [comment]
@jsparmani Please suggest if other routes are required

File: users/api/serializers.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+class ProfileUpdateSerializer(ModelSerializer):
+    class Meta:
+        model = Profile
+        fields = ['image', 'bio', 'year', 'branch']
+
+
+class ProfileListSerializer(ModelSerializer):
+    class Meta:
+        model = Profile
+        fields = ['user', 'image', 'bio', 'year', 'branch']
```
Comment (jsparmani) [inline users/api/serializers.py:21]: Why is the ListSerializer required? Is there any place where we would require a list of all profiles?

File: users/api/views.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    def perform_update(self, serializer):
+        payload = jwt_decoder(self.request.headers['Authorization'].split()[1])
+        serializer.save(author_id=payload['user_id'])
+
+
+class ProfileListAPIView(ListAPIView):
+    queryset = Profile.objects.all()
+    serializer_class = ProfileListSerializer
+    permission_classes = (AllowAny,)
```
Comment (jsparmani) [inline users/api/views.py:44]: Not sure if we require a `ListView` and `ListSerializer`. @Saurav-Shrivastav lemme know if you have any usecase about this in mind.

File: users/api/urls.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from django.urls import path
+
+from .views import ProfileDetailAPIView, ProfileListAPIView, ProfileUpdateAPIView
+
+urlpatterns = [
+    path('', ProfileListAPIView.as_view(), name='list'),
+    path('<user>', ProfileDetailAPIView.as_view(), name='detail'),
```
Comment (jsparmani) [inline users/api/urls.py:7]: append a / at the end of `<user>`

File: users/api/views.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    serializer_class = ProfileDetailSerializer
+    lookup_field = 'user'
+    permission_classes = (AllowAny,)
+
+
+class ProfileUpdateAPIView(RetrieveUpdateAPIView):
+    queryset = Profile.objects.all()
+    serializer_class = ProfileUpdateSerializer
+    lookup_field = 'user'
+    permission_classes = (IsAuthenticatedOrReadOnly, IsOwnerOrReadOnly)
```
Comment (jsparmani) [inline users/api/views.py:34]: I think this can be changed to `(IsAuthenticated, IsOwnerOrReadOnly)`

File: users/api/views.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+def jwt_decoder(encoded_token):
+    return jwt.decode(encoded_token, env('SIGNING_KEY'), algorithms=['HS256'])
+
+
+class ProfileDetailAPIView(RetrieveAPIView):
+    queryset = Profile.objects.all()
+    serializer_class = ProfileDetailSerializer
+    lookup_field = 'user'
+    permission_classes = (AllowAny,)
```
Comment (jsparmani) [inline users/api/views.py:27]: `DetailView` should only be available to logged in users

File: users/api/views.py (hunk 4 above)
Comment (Saurav-Shrivastav) [reply users/api/views.py:34]: Sure I'll get it changed

File: users/api/serializers.py (hunk 1 above)
Comment (Saurav-Shrivastav) [reply users/api/serializers.py:21]: Nope, I don't think so, will get it removed.

File: users/api/views.py (hunk 5 above)
Comment (Saurav-Shrivastav) [reply users/api/views.py:27]: Yes, on it!

File: users/api/urls.py (hunk 3 above)
Comment (Saurav-Shrivastav) [reply users/api/urls.py]: done

File: users/api/urls.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+from django.urls import path
+
+from .views import ProfileDetailAPIView, ProfileUpdateAPIView
+
+urlpatterns = [
+    path('profile/<user>/', ProfileDetailAPIView.as_view(), name='detail'),
+    path('profile/<user>/update/', ProfileUpdateAPIView.as_view(), name='update'),
```
Comment (jsparmani) [inline users/api/urls.py:7]: why is `<user>` required in the `update` call. It should be just `profile/update` that automatically takes the user which is authenticated.

File: users/api/serializers.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
+from rest_framework.serializers import ModelSerializer
+
+from users.models import Profile
+
+
+class ProfileDetailSerializer(ModelSerializer):
+    class Meta:
+        model = Profile
+        fields = ['user', 'image', 'bio', 'year', 'branch']
```
Comment (jsparmani) [inline users/api/serializers.py:9]: The `user` field in the `json` that's received should be another object having the details of user like _first_name_, _last_name_ and _email_id_.

File: users/api/urls.py (hunk 6 above)
Comment (Saurav-Shrivastav) [reply users/api/urls.py:7]: Yes I will fix this in a future commit!

**Comment by jsparmani** [comment]
@Saurav-Shrivastav There has to be some view my profile route also.

**Comment by Saurav-Shrivastav** [comment]
> @Saurav-Shrivastav There has to be some view my profile route also, that shows all the info of the user sending the request.

@jsparmani I figure that will be achieved with the ProfileDetailAPIView?

File: users/api/serializers.py (hunk 7 above)
Comment (Saurav-Shrivastav) [reply users/api/serializers.py]: Done

File: users/api/serializers.py [hunk 8]
```diff
@@ -1,9 +1,27 @@
...
+
+
+class UserSerializer(ModelSerializer):
+    class Meta:
+        model = MyUser
+        fields = [
+            'id',
+            'username',
+            'email',
+            'mobile',
```
Comment (jsparmani) [inline users/api/serializers.py:13]: don't send `mobile_no` for now

File: users/api/urls.py [hunk 9]
```diff
@@ -3,6 +3,6 @@
 from .views import ProfileDetailAPIView, ProfileUpdateAPIView
 
 urlpatterns = [
-    path('profile/<user>/', ProfileDetailAPIView.as_view(), name='detail'),
-    path('profile/<user>/update/', ProfileUpdateAPIView.as_view(), name='update'),
+    path('profile/', ProfileDetailAPIView.as_view(), name='detail'),
```
Comment (jsparmani) [inline users/api/urls.py:6]: add <user_id> support to this and create a new `/myprofile` route

File: users/api/serializers.py (hunk 8 above)
Comment (Saurav-Shrivastav) [reply users/api/serializers.py:24]: Done

File: users/api/urls.py (hunk 9 above)
Comment (Saurav-Shrivastav) [reply users/api/urls.py]: Completed

**Comment by jsparmani** [comment]
Okay, @Saurav-Shrivastav I will start to check this now

**Comment by Saurav-Shrivastav** [comment]
Cool cool

**Pull request merged**